library/alloc/src/boxed.rs

```diff
@@ -1043,6 +1042,35 @@ impl<T: Clone, A: Allocator + Clone> Clone for Box<T, A> {
     }
 }
 
+trait SpecClone {
+    fn spec_clone(&self) -> Self;
+}
```

Comment:
I don't think you need 3 separate `SpecClone` traits. All `SpecClone` implementations follow the form:

1. Create the uninitialized box (rc, arc)
2. Get a raw pointer `m` for the box, and clone the object into `m`
    * with `T: Clone`, do `m.write((**self).clone())`
    * with `T: Copy`, do `m.copy_from_nonoverlapping(&**self)`
3. Mark the box initialized with `assume_init()`

With this, only one trait is needed:

```rust
trait WriteCloneIntoRaw {
    unsafe fn write_clone_into_raw(&self, target: *mut Self);
}
impl<T: Clone> WriteCloneIntoRaw for T {
    default unsafe fn write_clone_into_raw(&self, target: *mut Self) {
        target.write(self.clone());
    }
}
impl<T: Copy> WriteCloneIntoRaw for T {
    unsafe fn write_clone_into_raw(&self, target: *mut Self) {
        target.copy_from_nonoverlapping(self, 1);
    }
}
```

and then the original SpecClone can be inlined directly into the `Clone`/`make_mut` impl:


```rust
impl<T: Clone, A: Allocator + Clone> Clone for Box<T, A> {
    fn clone(&self) -> Self {
        let mut boxed = Self::new_uninit_in(self.1.clone());
        unsafe {
            (*self).write_clone_into_raw(boxed.as_mut_ptr());
            boxed.assume_init()
        }
    }
}
```

Reply:
I've implemented this suggestion.